### PR TITLE
Fix inconsistencies in spec

### PR DESCRIPTION
**Pull request by adityasaky**

### CODE REVIEW

File: in-toto-spec.md [hunk 1]
```diff
@@ -529,106 +538,70 @@ To provide descriptive examples, we will adopt "canonical JSON," as described
...
-  ]
-}
-```
+### 4.2 General principles
 
-Where, ROLE is a dictionary whose `"_type"` field describes the metadata type (as
-described in sections 4.3 and 4.4). KEYID is the identifier of the key signing
-the ROLE dictionary. SIGNATURE is a signature of the canonical JSON form of
-ROLE.
+#### 4.2.1 Key formats
```
Comment (SantiagoTorres) [inline in-toto-spec.md:551]: I wonder if we want to mention that implementations may serialize to disk using whatever format for private keys they would like to use (I don't think this is clear, and tools like -golang and -rs opted for this to simplify the impl).

**Review by SantiagoTorres — CHANGES_REQUESTED** [review]
some minor nits. Thank you for the pass!

File: in-toto-spec.md [hunk 2]
```diff
@@ -229,7 +232,13 @@ trusted if present in the supply chain.
 In addition, the framework must provide an interface to interact with client
 systems that further verify the integrity of each step.
 
-### 1.6 Terminology
+### 1.6 in-toto Enhancements (ITEs)
+
+Enhancements to the in-toto specification can be submitted as ITEs.
```
[the remark this answers is not in the document]
Comment (adityasaky) [reply in-toto-spec.md:237]: Done

File: in-toto-spec.md [hunk 3]
```diff
@@ -529,106 +538,70 @@ To provide descriptive examples, we will adopt "canonical JSON," as described
 in
 [http://wiki.laptop.org/go/Canonical\_JSON](http://wiki.laptop.org/go/Canonical_JSON),
 as the data format. However, applications that desire to implement in-toto are
-not required to use JSON.  Discussion about the intended data format for
-in-toto can be found in the in-toto website.
-
-### 4.2 File formats: general principles
+not required to use JSON.
```
[the remark this answers is not in the document]
Comment (adityasaky) [reply in-toto-spec.md:541]: Made a minor edit to indicate multiple CJSON flavours exist.

File: in-toto-spec.md (hunk 1 above)
Comment (adityasaky) [reply in-toto-spec.md:552]: Added a note.

File: in-toto-spec.md [hunk 4]
```diff
@@ -529,106 +539,73 @@ To provide descriptive examples, we will adopt "canonical JSON," as described
...
-```
+| Method | KEYTYPE | SCHEME| KEYVAL | Notes |
+|--------|---------|-------|--------|-------|
+| RSASSA-PSS | rsa | rsassa-pss-sha256 | PEM formatted string | RSA keys must be at least 2048 bits |
+| ED25519 | ed25519 | ed25519 | 256 bit strings | |
+| ECDSA | ecdsa | ecdsa-sha2-nistp256 | 256 bit strings | |
 
-where PUBLIC and PRIVATE are both 32-byte (256-bit) strings.
+Note: This key format is for representation of keys in in-toto metadata only.
+Implementations may serialize keys to disk using any format they prefer.
```
Comment (lukpueh) [inline in-toto-spec.md:586]: Minor nit: Emphasizing that this is the key metadata format is a good idea. But I don't think we need to suggest, that in-toto users serialize keys to disk. With the new securesystemslib signer API they could be using HSMs or cloud signing providers (see https://github.com/in-toto/in-toto/issues/533, https://github.com/in-toto/in-toto/issues/532).

File: in-toto-spec.md [hunk 5]
```diff
@@ -529,106 +539,73 @@ To provide descriptive examples, we will adopt "canonical JSON," as described
...
-    "scheme" : "ecdsa-sha2-nistp256",
-    "keyval" : { "public" : "<PUBLIC>",
-                 "private" : "<PRIVATE>" }
-  }
-```
+| Method | KEYTYPE | SCHEME| KEYVAL | Notes |
+|--------|---------|-------|--------|-------|
+| RSASSA-PSS | rsa | rsassa-pss-sha256 | PEM formatted string | RSA keys must be at least 2048 bits |
+| ED25519 | ed25519 | ed25519 | 256 bit strings | |
+| ECDSA | ecdsa | ecdsa-sha2-nistp256 | 256 bit strings | |
```
[the remark this answers is not in the document]
Comment (lukpueh) [reply in-toto-spec.md:583]: That said, I also think it's a good idea to not require the `KEYVAL` dictionary to have the format it currently has. I do like the TUF phrasing "A dictionary containing the public portion of the key".

Just when we describe format of the currently supported keys, we should do it accurately.

File: in-toto-spec.md [hunk 6]
```diff
@@ -848,10 +823,17 @@ fields of a step and are applied sequentially on a queue of `"materials"` or
...
-list. By explicitly specifying `"DISALLOW *"`, in-toto verification fails if an
-artifact was not consumed by an earlier rule. Here, we describe an algorithm to
-illustrate the behavior of the rules being applied:
+subsequent rules.
+
+**NOTE:** There is an implicit `"ALLOW *"` at the end of each rule list. By
+explicitly specifying `"DISALLOW *"`, in-toto verification fails if an artifact
+was not consumed by an earlier rule. Not including `"DISALLOW *"` at the end of
+a rule list could allow artifacts to sneak into the step or inspection
+undetected. AS such, it is generally recommended that all rule lists include a
```
Comment (lukpueh) [inline in-toto-spec.md:832]:
```suggestion
undetected. As such, it is generally recommended that all rule lists include a
```

**Review by lukpueh — CHANGES_REQUESTED** [review]
Thanks for the patch, @adityasaky! I pointed out a couple of issues, most notably about key formats. The rest looks pretty good to me. 

I'd like to give the section about artifact rule processing another thorough pass and see if it really aligns with the reference implementation. But we can do this in a separate PR.

File: in-toto-spec.md [hunk 7]
```diff
@@ -529,106 +539,73 @@ To provide descriptive examples, we will adopt "canonical JSON," as described
...
-  { "keytype" : "ed25519",
-    "scheme" : "ed25519",
-    "keyval" : { "public" : "<PUBLIC>",
-                 "private" : "<PRIVATE>" }
-  }
-```
+KEYTYPE is a string denoting a public key signature system. SCHEME is a string
+denoting a corresponding signature scheme. KEYVAL is a dictionary containing the
+public portion of the key. The following table summarizes the expected entries
+for each signing algorithm supported by the reference implementation.
```
[the remark this answers is not in the document]
Comment (adityasaky) [reply in-toto-spec.md:577]: I'd prefer the specification to not specify this at all, tbh. Perhaps we should adopt POUFs for these matters. Do you reckon we describe the fields and leave it at that? Remove key schemes altogether and point them to implementation documentation for specifics?

File: in-toto-spec.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (adityasaky) [reply in-toto-spec.md:583]: > A dictionary containing the public portion of the key

This is actually the phrasing right above for KEYVAL, I'll fix the table.

File: in-toto-spec.md (hunk 4 above)
Comment (adityasaky) [reply in-toto-spec.md:586]: removed the second line altogether.

**Review by lukpueh — APPROVED** [review]

**Comment by lukpueh** [comment]
Thanks for the updates, @adityasaky! We can beautify JSON snippets in a separate PR.

**Pull request merged**